Add WENO helper functions

fmahebert opened the pull request:
## Proposed changes

Add helper functions that will be used in reconstructing the new WENO solution from previous local and neighbor data.

Some simplistic caching optimizations are made that will have to be generalized to use these functions on more general (non-uniform) grids.

### Types of changes:

- [ ] Bugfix
- [ ] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

fmahebert committed:
Add helper for WENO oscillation indicator

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.hpp:
```diff
@@ -0,0 +1,39 @@
+// mixed-derivative terms, we also count contributions from derivatives
+// (i.e., variations) in only one direction.
+//
+// Note: various references give closely-related oscillation / smoothness
```
I would like to see an `enhancement` issue to allow switching between the different OIs using an enum. I think it is at least somewhat interesting to compare.

fmahebert replied:
Opened issue #1437

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,145 @@
+  ModalVector deriv_coeffs(number_of_modes, 0.);
+  for (size_t i = 0; i < number_of_modes; ++i) {
+    for (size_t j = (i % 2 == 0 ? 1 : 0); j < i; j += 2) {
+      deriv_coeffs[j] += coeffs[i] * (2. * j + 1.);
```
Is this assuming a specific basis? If so a comment should be added and an ASSERT where the `Mesh` is available. I think this is only true for Legendre polynomials.

kidder replied:
not for this PR, but perhaps ModalVector should be templated on its basis?

nilsdeppe replied:
Yea, though that would be moving towards `DataMesh` :man_shrugging:

fmahebert replied:
If so, we should also template `DataVector`, because adding data from different grid points is just as meaningless as adding modes in different bases. (Am I just rephrasing what you are saying, @nilsdeppe?)

nilsdeppe replied:
Well I was asking @kidder what he thinks because I'm not sure what I think yet :P

kidder replied:
Never mind, I think the same reasons we decided not to template Mesh on the Basis will apply.
Perhaps the function should take the Basis and switch on it, with an ERROR for unimplemented

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,145 @@
+#include "Utilities/MakeArray.hpp"
+
+namespace {
+// Given the modal expansion corresponding to some function f(x), compute the
```
Is `x` the logical or inertial coordinate. This is important because if it's the logical coords then I worry about tensor densities and getting that right.

fmahebert commented:
Updated to remove entries that I had added to References.bib but wasn't citing from Doxygen, as this causes a documentation post-processing error. For now, these references are inlined in a C++ comment. In a later PR that adds the main WENO implementation, I will re-add the references and handle them more cleanly.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,147 @@
+
+// For a given pair of Legendre polynomial basis functions (P_m, P_n), compute
+// the sum of the integrals of all their derivatives:
+// \sum_{l=0}^{N} f(l) \int d^(l)/dx^(l) P_m(x) * d^(l)/dx^(l) P_n(x) * dx
```
what is `f(l)`?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,147 @@
+namespace {
+// Given the modal expansion corresponding to some function f(x), compute the
+// modal expansion corresponding to the derivative d/dx f(x).
+ModalVector modal_derivative(const ModalVector& coeffs) noexcept {
```
This seems like something that would be good to have available in `Spectral.hpp`/`cpp`

fmahebert replied:
Agreed. I feel it's outside the scope of this PR though... would you be okay with me opening an issue to remind us to revisit this in the future?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,147 @@
+// The indicator matrix is computed by tensor product, using the tensor-product
+// nature of the Legendre basis in N dimensions.
+//
+// Note that we make a change (correction?) to the reference. In each direction
```
I'm not sure this is a correction. Have you compared if there is any difference between the two approaches? To me starting at 1 sounds better since the 0 is just a constant term, no? As far as I can tell from the paper and other references the zeroth derivative term is intentionally omitted.

fmahebert replied:
The paper suggests taking the tensor product of, for example: (0, d/dx, d2/dx2) cross (0, d/dy, d2/dy2). This produces, in 2D, the terms: d2/(dx dy), d3/(dx2 dy), d3/(dx dy2), d4/(dx2 dy2). Note this does not produce terms d/dx, d2/dx2, d/dy, d2/dy2.

I instead compute: (1, d/dx, d2/dx2) cross (1, d/dy, d2/dy2) – 1.  This produces the terms listed above including the missing terms.

The zeroth derivative in _all_ directions is indeed still omitted, but I don’t see why terms with a zeroth derivative in _a_ direction should also be omitted.

nilsdeppe replied:
Yes, I agree that only the total zeroth derivative should be omitted. I'm not sure exactly what should be added to the comment to make it clearer, but something should be, I think :)

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,147 @@
+             make_array<VolumeDim>(Spectral::Quadrature::GaussLobatto),
+         "No implementation for mesh: " << mesh);
+
+  // TODO(): Generalize the caching infrastructure to handle multiple different
```
`TODO` -> issue unless you're planning to do this before it gets merged.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,147 @@
+  const ModalVector coeffs = to_modal_coefficients(data, mesh);
+
+  double result = 0.;
+  for (size_t m = 0; m < mesh.number_of_grid_points(); ++m) {
```
This is also inconsistent with the paper as far as I can tell. It should start at 1, not 0.

fmahebert replied:
I do agree this could start at 1; "physically", the mean doesn't contribute any oscillation, so the 0'th spectral coeff shouldn't play into the indicator.

I've changed the code to compute an (N-1)^2 indicator matrix and contract it with (N-1) modal coeffs (the numerical results are unchanged within the precision). This is less elegant, but may provide a tiny performance benefit.

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,147 @@
+  return result;
+}
+
+// Compute the indicator matrix, as given by Eq. 28 of Dumbser2007.
```
I don't see an indicator matrix in Eq. 28 of `10.1016/j.jcp.2006.06.043`, I think it's eq. 23 that you mean.

fmahebert replied:
So it turns out I was reading a preprint, and the preprint has some notable differences from the published version. This is why some things must have seemed inconsistent. Going through and updating as necessary now...

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.hpp:
```diff
@@ -4,13 +4,28 @@
 #pragma once
 
 #include <cstddef>
+#include <unordered_map>
```
Can you make the commit message more informative please?

fmahebert replied:
You haven't told me what you want clarified. Would it be helpful to add a sentence like "Adds a helper function that computes the WENO reconstruction as the weighted sum of different input polynomials" to the commit message?

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.hpp:
```diff
@@ -57,5 +72,86 @@ template <size_t VolumeDim>
+    auto& local_polynomial = (*local_tensor)[tensor_storage_index];
+
+    // Store linear weights in `local_weights` and `neighbor_weights`
+    // TODO(): Generalize to handle weights for multiple neighbors per face.
```
`TODO` -> issue or do it.

fmahebert replied:
Opened issue #1436

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.hpp:
```diff
@@ -57,5 +72,86 @@ template <size_t VolumeDim>
+
+    // Perform reconstruction, by superposition of local and neighbor
+    // polynomials.
+    // TODO(): See whether mean_value computations can be avoided. Note that:
```
again, issue or do it.

fmahebert replied:
Opened issue #1437

nilsdeppe commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.hpp:
```diff
@@ -57,5 +72,86 @@ template <size_t VolumeDim>
+        Variables<TagsList>,
+        boost::hash<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>>&
+        neighbor_vars) noexcept {
+  // Pre-allocate storage for the weights
```
I don't understand, where's the memory being allocated?

fmahebert replied:
This was a bad choice of words on my part to describe premature, un-measured "optimization" :-/. I've moved the unordered_map back to a more restricted scope inside the loops.

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,215 @@ void test_oscillation_indicator_3d() noexcept {
+
+void test_reconstruction_1d() noexcept {
+  const double neighbor_linear_weight = 0.001;
+  const auto mesh = Mesh<1>({{5}}, Spectral::Basis::Legendre,
```
make `5` a variable instead of a magic number used in several places (`constexpr size_t`)

fmahebert replied:
I don't like making this a variable, because that suggests (imo) that changing its value is a good idea, whereas here the entire test function would need to be adapted for a different resolution.

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,215 @@ void test_oscillation_indicator_3d() noexcept {
+
+void test_reconstruction_1d() noexcept {
+  const double neighbor_linear_weight = 0.001;
+  const auto mesh = Mesh<1>({{5}}, Spectral::Basis::Legendre,
```
Why not `const Mesh<1> mesh{{...}};`? The `auto` here seems a bit silly

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,215 @@ void test_oscillation_indicator_3d() noexcept {
+                            Spectral::Quadrature::GaussLobatto);
+  const auto coords = logical_coordinates(mesh);
+
+  const auto evaluate = [&coords](
```
maybe something like `evaulate_expected_result` or something? Just `evaluate` seems comparable to naming functions `function`

fmahebert replied:
This is used to evaluate the input as well the expected result. Would you accept `evaluate_polynomial` ?

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,215 @@ void test_oscillation_indicator_3d() noexcept {
+  SlopeLimiters::Weno_detail::reconstruct_from_weighted_sum<ScalarTag>(
+      make_not_null(&scalar), mesh, neighbor_linear_weight, neighbor_vars);
+
+  // Expected result computed in Mathematica by computing oscillation indicator
```
Why not do this with calling into python rather than some Mathematica code that I can't check to see if you did it correctly and isn't in the repo? One of the main goals of having `pypp` was to support computing the second result inside the code rather than in something proprietary like Mathematica.

fmahebert replied:
Will think about this one... it is not obvious to me if it is reasonably straightforward to compute spectral derivatives and transformations in our Python env, or alternatively, if it is reasonably straightforward to pass the sets of extra arguments though `pypp` in the first place.

fmahebert replied:
It was not straightforward to test this in Python, for several reasons:
- the complicated datastructure associated with input data (a map of tensors)
- either the oscillation indicator needs to be computed in Python, which requires spectral transforms or differentiations, or the indicator for each tensor component for each neighbor needs to be passed in as an extra argument (which isn't supported).

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,215 @@ void test_oscillation_indicator_3d() noexcept {
+  CHECK_ITERABLE_APPROX(scalar, expected_reconstructed_scalar);
+}
+
+void test_reconstruction_2d() noexcept {
```
Basically the same comments for 2d and 3d.

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,215 @@ void test_oscillation_indicator_3d() noexcept {
+};
+
+void test_reconstruction_1d() noexcept {
+  const double neighbor_linear_weight = 0.001;
```
Is there a good reason not to test with another value of `neighbor_linear_weight`?

fmahebert replied:
I've changed the 1D case to use a different weight.

nilsdeppe commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,215 @@ void test_oscillation_indicator_3d() noexcept {
+
+void test_reconstruction_1d() noexcept {
+  const double neighbor_linear_weight = 0.001;
+  const auto mesh = Mesh<1>({{5}}, Spectral::Basis::Legendre,
```
Is testing with different number of grid points not possible? I realize in 3d this might be too expensive, but in 1d and probably 2d it should be fine.

fmahebert replied:
Possible, but requires a new TEST_CASE for each resolution, for the same caching-related reasons given to justify why these TEST_CASES aren't merged.

fmahebert commented:
@nilsdeppe I've added fixups that essentially rewrite the oscillation indicator documentation, which I hope will clear up some of the confusion about what is being done. Let me know if this is clearer!

I believe I've addressed most of your comments, though I am looking for clarification on a few comments, and also I still need to think about the feasibility of a Python test.

nilsdeppe commented:
Looks good to me. Please rebase and squash :)

fmahebert committed:
Add helper for WENO reconstruction

Adds a helper function that computes the WENO reconstruction as
the weighted sum of different input polynomials.

fmahebert commented:
Rebased and squashed

nilsdeppe commented:
Restarted all builds to make sure recent merges don't interfere. Previous travis failure was a timeout of clang release.

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,175 @@
+//
+// Note that we compute the product using derivatives 0...N in each dimension.
+// We then subtract off the term with 0 derivatives in all dimensions, because
+// that term is just the original data and should be ommitted.
```
omitted

kidder commented on src/Evolution/DiscontinuousGalerkin/SlopeLimiters/WenoHelpers.cpp:
```diff
@@ -0,0 +1,175 @@
+  // once at the beginning of the simulation. This caching will have to be
+  // generalized to handle simulations with more than one mesh.
+  static const Matrix indicator_matrix = compute_indicator_matrix(mesh);
+
```
Please then add an ASSERT that `indicator_matrix` is compatible with `mesh` so we don't forget this when there is more than one mesh

kidder commented on tests/Unit/Evolution/DiscontinuousGalerkin/SlopeLimiters/Test_WenoHelpers.cpp:
```diff
@@ -82,11 +97,217 @@ void test_oscillation_indicator_3d() noexcept {
+  const auto coords = logical_coordinates(mesh);
+
+  // 3D case has so many modes... so we simplify by only setting 6 of them, the
+  // choice if modes to use here is arbitrary.
```
if -> of

fmahebert commented:
@kidder fixups posted

kidder reviewed (commented):
I am happy with the fixups

fmahebert commented:
Squashed

kidder commented:
@fmahebert this has conflicts with what I just merged...

fmahebert commented:
Yes, I didn't know which one would be merged first so I didn't want to add dependencies :P. Rebased now.
